File: implementations/java-springboot/examples/src/test/java/com/example/accountservice/account/application/command/TransferServiceTest.java
package com.example.accountservice.account.application.command;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.example.accountservice.account.domain.Account;
import com.example.accountservice.account.domain.AccountException;
import com.example.accountservice.account.domain.AccountFindQuery;
import com.example.accountservice.account.domain.AccountRepository;
import com.example.accountservice.account.domain.AccountsWithCount;
import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

// TransferEligibilityService (a Domain Service) is a plain class, so it is not mocked here — this
// spec verifies that the Application layer loads both accounts, delegates to the actual
// eligibility logic, and, once approved, saves both accounts in a single transaction
// (saveAccounts).
@ExtendWith(MockitoExtension.class)
class TransferServiceTest {

    @Mock private AccountRepository accountRepository;

    private TransferService service;

    @BeforeEach
    void setUp() {
        service = new TransferService(accountRepository);
    }

    @Test
    void saves_both_accounts_in_one_transaction_when_approved() {
        Account source = Account.create("owner-1", "owner-1@example.com", "KRW");
        source.deposit(10000);
        Account target = Account.create("owner-2", "owner-2@example.com", "KRW");
        when(accountRepository.findAccounts(
                        new AccountFindQuery(0, 1, source.getAccountId(), "owner-1", null)))
                .thenReturn(new AccountsWithCount(List.of(source), 1));
        when(accountRepository.findAccounts(
                        new AccountFindQuery(0, 1, target.getAccountId(), null, null)))
                .thenReturn(new AccountsWithCount(List.of(target), 1));

        TransferResult result =
                service.transfer(
                        new TransferCommand(
                                source.getAccountId(), target.getAccountId(), "owner-1", 4000));

        assertThat(result.sourceTransaction().type()).isEqualTo("WITHDRAWAL");
        assertThat(result.targetTransaction().type()).isEqualTo("DEPOSIT");
        assertThat(source.getBalance().amount()).isEqualTo(6000);
        assertThat(target.getBalance().amount()).isEqualTo(4000);
        verify(accountRepository).saveAccounts(source, target);
    }

    @Test
    void throws_exception_and_does_not_save_when_source_account_is_not_found() {
        when(accountRepository.findAccounts(new AccountFindQuery(0, 1, "missing", "owner-1", null)))
                .thenReturn(new AccountsWithCount(List.of(), 0));

        assertThatThrownBy(
                        () ->
                                service.transfer(
                                        new TransferCommand(
                                                "missing", "account-2", "owner-1", 1000)))
                .isInstanceOf(AccountException.class)
                .extracting(e -> ((AccountException) e).code())
                .isEqualTo(AccountException.ErrorCode.ACCOUNT_NOT_FOUND);
        verify(accountRepository, never())
                .saveAccounts(
                        org.mockito.ArgumentMatchers.any(), org.mockito.ArgumentMatchers.any());
    }

    @Test
    void throws_exception_and_does_not_save_when_balance_is_insufficient() {
        Account source = Account.create("owner-1", "owner-1@example.com", "KRW");
        source.deposit(1000);
        Account target = Account.create("owner-2", "owner-2@example.com", "KRW");
        when(accountRepository.findAccounts(
                        new AccountFindQuery(0, 1, source.getAccountId(), "owner-1", null)))
                .thenReturn(new AccountsWithCount(List.of(source), 1));
        when(accountRepository.findAccounts(
                        new AccountFindQuery(0, 1, target.getAccountId(), null, null)))
                .thenReturn(new AccountsWithCount(List.of(target), 1));

        assertThatThrownBy(
                        () ->
                                service.transfer(
                                        new TransferCommand(
                                                source.getAccountId(),
                                                target.getAccountId(),
                                                "owner-1",
                                                5000)))
                .isInstanceOf(AccountException.class)
                .extracting(e -> ((AccountException) e).code())
                .isEqualTo(AccountException.ErrorCode.INSUFFICIENT_BALANCE);
        verify(accountRepository, never())
                .saveAccounts(
                        org.mockito.ArgumentMatchers.any(), org.mockito.ArgumentMatchers.any());
    }
}

File: implementations/kotlin-springboot/examples/src/test/kotlin/com/example/accountservice/account/application/command/TransferServiceTest.kt
package com.example.accountservice.account.application.command

import com.example.accountservice.account.domain.Account
import com.example.accountservice.account.domain.AccountFindQuery
import com.example.accountservice.account.domain.AccountNotFoundException
import com.example.accountservice.account.domain.AccountRepository
import com.example.accountservice.account.domain.InsufficientBalanceException
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

// TransferEligibilityService (a Domain Service) is a plain class, so it is not mocked — this spec
// verifies the flow where the Application layer loads both accounts, delegates to the real decision
// logic, and, when approved, saves both accounts in a single transaction (saveAccounts).
class TransferServiceTest {
    private val accountRepository = mockk<AccountRepository>(relaxed = true)
    private val service = TransferService(accountRepository)

    @Test
    fun `saves both accounts in a single transaction when approved`() {
        val source = Account.create("owner-1", "KRW", "owner-1@example.com")
        source.deposit(10000)
        val target = Account.create("owner-2", "KRW", "owner-2@example.com")
        every {
            accountRepository.findAccounts(AccountFindQuery(page = 0, take = 1, accountId = source.accountId, ownerId = "owner-1"))
        } returns (listOf(source) to 1L)
        every {
            accountRepository.findAccounts(AccountFindQuery(page = 0, take = 1, accountId = target.accountId))
        } returns (listOf(target) to 1L)

        val result = service.transfer(TransferCommand(source.accountId, target.accountId, "owner-1", 4000))

        assertThat(result.sourceTransaction.type).isEqualTo("WITHDRAWAL")
        assertThat(result.targetTransaction.type).isEqualTo("DEPOSIT")
        assertThat(source.balance.amount).isEqualTo(6000)
        assertThat(target.balance.amount).isEqualTo(4000)
        verify(exactly = 1) { accountRepository.saveAccounts(source, target) }
    }

    @Test
    fun `throws an exception and does not save when the withdrawal account cannot be found`() {
        every {
            accountRepository.findAccounts(AccountFindQuery(page = 0, take = 1, accountId = "missing", ownerId = "owner-1"))
        } returns (emptyList<Account>() to 0L)

        assertThrows<AccountNotFoundException> {
            service.transfer(TransferCommand("missing", "account-2", "owner-1", 1000))
        }
        verify(exactly = 0) { accountRepository.saveAccounts(any(), any()) }
    }

    @Test
    fun `throws an exception and does not save when the balance is insufficient`() {
        val source = Account.create("owner-1", "KRW", "owner-1@example.com")
        source.deposit(1000)
        val target = Account.create("owner-2", "KRW", "owner-2@example.com")
        every {
            accountRepository.findAccounts(AccountFindQuery(page = 0, take = 1, accountId = source.accountId, ownerId = "owner-1"))
        } returns (listOf(source) to 1L)
        every {
            accountRepository.findAccounts(AccountFindQuery(page = 0, take = 1, accountId = target.accountId))
        } returns (listOf(target) to 1L)

        assertThrows<InsufficientBalanceException> {
            service.transfer(TransferCommand(source.accountId, target.accountId, "owner-1", 5000))
        }
        verify(exactly = 0) { accountRepository.saveAccounts(any(), any()) }
    }
}
